pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -3078,6 +3078,24 @@ public void revertReplaceSegments(String tableNameWithType, String segmentLineag
           throw new RuntimeException(errorMsg);
         }
 
+        // We do not allow to revert the lineage entry which segments in 'segmentsTo' appear in 'segmentsFrom' of other
```

Comment:
I think that we are making the assumption that the snapshots can be at most 2. Let's add `TODO` comment here that we need to expand this logic when we support N snapshots.

Reply:
Done